src/commcare_cloud/commands/terraform/aws.py

```diff
@@ -287,13 +289,108 @@ class AwsSignIn(CommandBase):
+
+
+@memoized
+def _aws_sign_in_with_sso(environment):
+    """
+    Create a temp session through MFA for a given aws profile
+
+    :param aws_profile: The name of an existing aws profile to create a temp session for
+    :param duration_minutes: How long to set the session expiration if a new one is created
+    :param force_new: If set to True, creates new credentials even if valid ones are found
+    :return: The name of temp session profile.
+             (Always the passed in profile followed by ':session')
+    """
+    aws_session_profile = '{}:session'.format(environment.terraform_config.aws_profile)
+    # todo: add `... or if _date_modified(aws_config_path) > _date_modified(aws_credentials_path)`
+    if not _has_profile_for_sso(aws_session_profile):
+        puts(color_notice("Configuring SSO. To further customize, run `aws configure sso --profile {}`".format(aws_session_profile)))
+        _write_profile_for_sso(
+            aws_session_profile,
+            sso_start_url=environment.aws_config.sso_config.sso_start_url,
+            sso_account_id=environment.aws_config.sso_config.sso_account_id,
+            sso_region=environment.aws_config.sso_config.sso_region,
+            region=environment.aws_config.sso_config.region,
+        )
+
+    if not _has_valid_session_credentials_for_sso():
+        check_output(['aws', 'sso', 'login'], env={'AWS_PROFILE': aws_session_profile})
+
+    if not _has_valid_v1_session_credentials(aws_session_profile):
+        caller_identity = _get_caller_identity({'AWS_PROFILE': aws_session_profile})
+        if not caller_identity:
+            raise ValueError((
+                "SSO profile mis-configured and cannot fix automatically. "
+                "Edit [profile {}] in ~/.aws/config manually;"
+                "to start over, remove it from the file manually and try again.").format(aws_session_profile))
+        # Until this is built in to aws, we need this insane workaround to get backwards compatibility
+        # with the credential format that terraform uses
+        cli_cache_dir = os.path.expanduser('~/.aws/cli/cache/')
+
+        for filename in os.listdir(cli_cache_dir):
+            with open(os.path.join(cli_cache_dir, filename)) as f:
+                try:
+                    data = json.load(f)
+                except JSONDecodeError:
+                    continue
+                else:
+                    if data.get('ProviderType') != 'sso':
+                        continue
+                    credentials = data['Credentials']
+                    comparison = _get_caller_identity({
+                        'AWS_ACCESS_KEY_ID': credentials.get('AccessKeyId'),
+                        'AWS_SECRET_ACCESS_KEY': credentials.get('SecretAccessKey'),
+                        'AWS_SESSION_TOKEN': credentials.get('SessionToken'),
+                    })
+                    if comparison == caller_identity:
+                        break
```

Comment:
Consider moving this out to a helper function since it's a rather big "insane workaround."

Also, the `else` (`raise ValueError...`) clause below will not be executed if this `for`-loop exits without finding credentials. Is that a bug?

Reply:
Yeah good point I'll definitely move this out to a helper function. Have to admit I did this a bit hastily so after applying your feedback I'll also do another quick pass myself at tidying up / tying up loose ends

Reply:
I don't think I understand the comment about the `else:` (`raise ValueError(...)`) below. The way I'm intending to use it is as part of a `for/else`, where the `else` is called if the end of the `for` loop is reached without hitting a `break`.

I found a way to contrive a test of this (stuck a pause in after the first call to `_get_caller_identity` and then snuck in and manually deleted all the files in `~/.aws/cli/cache/`), and it raised this error the way I expected it to

Reply:
+1 for the test. Yay!

Oh, I misunderstood the code. I thought the `else` was the other branch of `if not _has_valid_v1_session_credentials(aws_session_profile):`. Now I see it's with the `for`-loop. All good.

Reply:
It was just a manual test, so I think I don't deserve the "yay" 😬, but it was just enough to make sure that my understanding of the code logic was on the right track